### ISSUE TITLE
highlight syntax for functional languages is ugly, but in native pygments is ok

### ISSUE THREAD

**Issue by kwladyka** [gohugoio/hugo#6068]
Example of code:
```clojure
(def counter (atom 0))
(def foo (atom 1))
(defn slow-inc [n]
  (swap! counter inc)
  (Thread/sleep 200)
  (inc n))

(pmap
  (fn [_]
    (swap! foo slow-inc))
  (range 100))

@counter
@foo
```

Wiht https://highlightjs.org/ and `darkula` looks like this:
![image](https://user-images.githubusercontent.com/3903726/60382687-2388fd00-9a67-11e9-9e2c-95e1e18d3a9d.png)

With http://pygments.org/ and for example `perldoc`
![image](https://user-images.githubusercontent.com/3903726/60382696-45827f80-9a67-11e9-8ef4-8ce900eb852e.png)

while in hugo with `perldoc` like this:
![image](https://user-images.githubusercontent.com/3903726/60382699-6a76f280-9a67-11e9-9e5a-f4b5e185ff6a.png)

with `pygmentsUseClassic: "true"` and `pygmentsUseClassic: "false"`

Why there are so huge differences? hugo make highlighting really bad. How to fix it?

**Comment by kwladyka** [comment]
Anybody? How to fix it? Is it a bug?

**Comment by kwladyka** [comment]
Generated Hugo HTML:

```html
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-clojure" data-lang="clojure">(<span style="color:#66d9ef">def </span>counter (atom <span style="color:#ae81ff">0</span>))
(<span style="color:#66d9ef">def </span>foo (atom <span style="color:#ae81ff">1</span>))
(<span style="color:#66d9ef">defn </span>slow-inc [n]
  (swap! counter inc)
  (Thread/sleep <span style="color:#ae81ff">200</span>)
  (inc n))

(pmap
  (<span style="color:#66d9ef">fn </span>[_]
    (swap! foo slow-inc))
  (range <span style="color:#ae81ff">100</span>))

<span style="color:#f92672">@</span>counter
<span style="color:#f92672">@</span>foo
=&gt; <span style="color:#ae81ff">1602</span>
=&gt; <span style="color:#ae81ff">101</span></code></pre></div>
```

**Comment by kwladyka** [comment]
https://discourse.gohugo.io/t/bug-highlight-syntax-for-functional-languages-is-wrong-but-in-native-pygments-is-ok/19569

Please also see topic here

**Comment by moorereason** [comment]
Hugo's built-in highlighting uses [Chroma](https://github.com/alecthomas/chroma).  I don't speak clojure, but I'm assuming this is an issue with the Chroma [clojure lexer](https://github.com/alecthomas/chroma/blob/0c0b382eca61a71c1eb4cb4dea2bc78aa4939d96/lexers/c/clojure.go).  Please raise an issue upstream in the Chroma project.

**Comment by alecthomas** [comment]
As stated in the linked issue, I can't replicate this with the chroma command-line tool:

```
$ cat t.clj
(def counter (atom 0))
(def foo (atom 1))
(defn slow-inc [n]
  (swap! counter inc)
  (Thread/sleep 200)
  (inc n))

(pmap
  (fn [_]
    (swap! foo slow-inc))
  (range 100))

@counter
@foo
$ chroma --html-only --html ./t.clj
<pre class="chroma"><span class="p">(</span><span class="k">def </span><span class="nv">counter</span> <span class="p">(</span><span class="nf">atom</span> <span class="mi">0</span><span class="p">))</span>
<span class="p">(</span><span class="k">def </span><span class="nv">foo</span> <span class="p">(</span><span class="nf">atom</span> <span class="mi">1</span><span class="p">))</span>
<span class="p">(</span><span class="kd">defn </span><span class="nv">slow-inc</span> <span class="p">[</span><span class="nv">n</span><span class="p">]</span>
  <span class="p">(</span><span class="nf">swap!</span> <span class="nv">counter</span> <span class="nv">inc</span><span class="p">)</span>
  <span class="p">(</span><span class="nf">Thread/sleep</span> <span class="mi">200</span><span class="p">)</span>
  <span class="p">(</span><span class="nb">inc </span><span class="nv">n</span><span class="p">))</span>

<span class="p">(</span><span class="nf">pmap</span>
  <span class="p">(</span><span class="k">fn </span><span class="p">[</span><span class="nv">_</span><span class="p">]</span>
    <span class="p">(</span><span class="nf">swap!</span> <span class="nv">foo</span> <span class="nv">slow-inc</span><span class="p">))</span>
  <span class="p">(</span><span class="nb">range </span><span class="mi">100</span><span class="p">))</span>

<span class="o">@</span><span class="nv">counter</span>
<span class="o">@</span><span class="nv">foo</span>
</pre>
```

This HTML looks identical to Pygments and nothing like the HTML from Hugo.

**Comment by kwladyka** [comment]
@moorereason Are you sure it is Chroma issue?

From what @alecthomas pasted it looks like Chroma works well.

**Comment by kwladyka** [comment]
Ok guys it is my last issue moving blog to hugo. Any hints what else I can do? Or it is just a Hugo bug and I can do nothing.
@moorereason any suggestion?

**Comment by moorereason** [comment]
@kwladyka, to work around this issue, you need to force the use of CSS classes instead of inline styles with the `pygmentsUseClasses` option.

> Should we remove `upstream` label?

No.  I have updated the upstream issue with a reproduction of the issue outside of Hugo.

**Comment by alecthomas** [comment]
Reading back, I honestly can't even tell what the issue is here.

The original complaint is:

> with pygmentsUseClassic: "true" and pygmentsUseClassic: "false"
> ...
> Why there are so huge differences? hugo make highlighting really bad. How to fix it?

It seems like the complaint is just that Hugo's syntax highlighting is "ugly", using either Chroma or Pygments. Given that Chroma this Clojure fragment identically to Pygments:

![image](https://user-images.githubusercontent.com/41767/61271754-ca4ff600-a7e8-11e9-9f05-bc2b3b2ab5b6.png)

(from the newly created [Chroma Playground](https://swapoff.org/chroma/playground/))

I don't know what else to say.

**Comment by kwladyka** [comment]
@alecthomas Not sure what you mean.

Here I described WHAT is wrong:
https://github.com/gohugoio/hugo/issues/6068#issue-462288805

Here I described WHY it is wrong:
https://github.com/gohugoio/hugo/issues/6068#issuecomment-510023145

Here is deeper WHY by @moorereason :
https://github.com/alecthomas/chroma/issues/259#issuecomment-511537882

**Comment by kwladyka** [comment]
>  to work around this issue, you need to force the use of CSS classes instead of inline styles with the pygmentsUseClasses option.

@moorereason I have right HTML, but still wrong highlight. I mean words are wrapped like they should (like in pygments) this time, but highlight is still wrong.

More details:
https://github.com/alecthomas/chroma/issues/259#issuecomment-511736512

**Comment by alecthomas** [comment]
Is the image I just posted right or wrong?

**Comment by kwladyka** [comment]
@alecthomas it is the right one

**Comment by alecthomas** [comment]
Okay so that is generated directly by Chroma (try it yourself on the [Chroma Playground](https://swapoff.org/chroma/playground/)) exactly as Pygments does. This suggests that if Hugo does not look correct it is either a) misconfigured or b) has a bug. Either way, this is outside my area of expertise.

**Comment by kwladyka** [comment]
I can confirm https://swapoff.org/chroma/playground/ show good highlight.
@alecthomas thank you for your time

So back again to Hugo. Oh man, so confusing. Does Hugo use the latest ver. of Chroma? To be honest I can't provide more useful details. We need Hugo expert to figure out this.

**Comment by moorereason** [comment]
@kwladyka,
It appears to me that Hugo and Chroma are indeed working correctly.  I incorrectly conflated two items in this discussion that sent me off in the wrong direction.  I'm able to make everything work with the following config.toml settings:

```yaml
pygmentsCodeFences: true
pygmentsStyle: "perldoc"
```

@alecthomas,
Sorry for the noise, but thank you for being patient with us as we worked through this.

**Comment by kwladyka** [comment]
@moorereason hmm it doesn't work for me. I am doing this like that from the beginning.

What is your hugo version?
```
hugo version
Hugo Static Site Generator v0.55.6/extended darwin/amd64 BuildDate: unknown
```

**Comment by moorereason** [comment]
You're correct, @kwladyka.  (And I keep being wrong in this thread!)

This chroma bug was fixed upstream in https://github.com/alecthomas/chroma/commit/ea14dd8660c608784452d795ad4f0f2373a4aba8.  We need to update the chroma dependency to v0.6.4.

Cc: @bep

**Comment by moorereason** [comment]
Fixed in 95b1d3013b4717f8b02093a99d1d0c4a6a1ca929.

**Comment by kwladyka** [comment]
@moorereason  Do you know when will be release with this fix?
